### PR TITLE
feature: modules/machine_learning 

### PR DESCRIPTION
**Pull request by Zephreo**
## Adds a basic ML Neural Networking library (`machine_learning.cpp`, `machine_learning.h`)

- can be used to make NN models with variable layers in sequential order
- Each layer can have a unique activation function and node count
  - Currently only fully connected (Dense) layers with 1 dimensional IO supported 
  - Currently only Sigmoid and ReLu activation functions supported
- Extra Loss functions can be easily added later on
  - Currently only Mean Squared Error (MSE) implemented 
- Testing can be found in `test_machine_learning.cpp`, no unit tests yet

## Overhauled matrix_2d library (`matrix_2d.cpp`, `matrix_2d.h`)

- Fully backwards compatible with old code (i.e. will not break existing code that uses matrix_2d)
- Variable sizing supported (previously only 3x3 matrices)
  - Uses heap memory with custom destructor/assignment operators for scoped memory management
- Added `matrix_slice()`, `matrix_transpose()`, `matrix_horizontal_concat()`, `matrix_add()`, and `matrix_multiply_components()` functions
- Overloaded many operators for easier element wise operations.
  - e.g. inequality < or > returns a matrix with 1 if true and 0 if false for each cell 
- iterator for c++ enhanced for loops, rows first then columns (left to right, top to bottom)
- new unit tests and regular test for testing the new functionality. (`test_matrix_2d.cpp`, `unit_test_matrix.cpp`)
## new Game AI API (`game_learning.cpp` and `game_learning.h`)
- A `Game` class has been added, any Game that inherits from this class and implements its functions can use the Game API
- Currently 3 Agents to choose from
  - `RandomAgent`
    - Plays random moves
  - `MinimaxAgent`
    - Uses the minimax algorithm to find the optimal move to make
     - Very Slow only usable on small games with few possible moves and few moves per game 
    - Caches results to be faster
  - `QAgent`
    - Uses the QLearning algorithm to converge to the optimal move to make
    - Requires training using the `agent.train()` method before hand
    - Stores a big table mapping every possible game state to every possible action
    - Plays simulated games against itself
      - When it wins it rewards the moves it made so as to increase the probability of choosing those moves again
      - When it loses it punishes the moves it made so as to decrease the probability of choosing those moves again
- Uses dynamic systems so as to allow any possible game to implement the API while still providing the AI with machine readable states 
- Tested in `test_machine_learning.cpp` currently only `TicTacToe` and `Pong` games have been tested on the system.

### CODE REVIEW

**Review by timmydzeko — COMMENTED** [review]
Looks great! Will get @Alex-MobileApps to review as well seeing as there are so many files. But haven't spotted anything that needs changing or updating.

**Comment by Alex-MobileApps** [comment]
Hi Harry, on the Mac M1 chip I am receiving the following issues:
- draw game score in pong is creating a splash kit assertion error (please see attached screenshots, this doesn't occur in the first game but it does in all following games)
- Progress bar in the demo is only showing at 100% and not throughout training for the pong qlearning agent
Screen Shot 2022-09-03 at 1 19 26 pm
Screen Shot 2022-09-03 at 1 20 46 pm

Other than that, everything is looking good

**Review by Alex-MobileApps — APPROVED** [review]
Looks good to me now Harry
